Address safer C++ static analysis warnings in WKWebViewConfiguration

cdumez opened the pull request:
#### c3f3b7cda7e892c4109b6c68ce2871ff12b34d63
<pre>
Address safer C++ static analysis warnings in WKWebViewConfiguration
<a href="https://bugs.webkit.org/show_bug.cgi?id=288040">https://bugs.webkit.org/show_bug.cgi?id=288040</a>

Reviewed by Timothy Hatcher and Darin Adler.

* Source/WebKit/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration dealloc]):
(-[WKWebViewConfiguration _protectedPageConfiguration]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration processPool]):
(-[WKWebViewConfiguration setProcessPool:]):
(-[WKWebViewConfiguration preferences]):
(-[WKWebViewConfiguration setPreferences:]):
(-[WKWebViewConfiguration userContentController]):
(-[WKWebViewConfiguration setUserContentController:]):
(-[WKWebViewConfiguration _requiredWebExtensionBaseURL]):
(-[WKWebViewConfiguration _setRequiredWebExtensionBaseURL:]):
(-[WKWebViewConfiguration _strongWebExtensionController]):
(-[WKWebViewConfiguration _weakWebExtensionController]):
(-[WKWebViewConfiguration _setWeakWebExtensionController:]):
(-[WKWebViewConfiguration setWebExtensionController:]):
(-[WKWebViewConfiguration websiteDataStore]):
(-[WKWebViewConfiguration setWebsiteDataStore:]):
(-[WKWebViewConfiguration defaultWebpagePreferences]):
(-[WKWebViewConfiguration setDefaultWebpagePreferences:]):
(-[WKWebViewConfiguration _visitedLinkStore]):
(-[WKWebViewConfiguration _setVisitedLinkStore:]):
(-[WKWebViewConfiguration setURLSchemeHandler:forURLScheme:]):
(-[WKWebViewConfiguration urlSchemeHandlerForURLScheme:]):
(-[WKWebViewConfiguration _relatedWebView]):
(-[WKWebViewConfiguration _webViewToCloneSessionStorageFrom]):
(-[WKWebViewConfiguration _setWebViewToCloneSessionStorageFrom:]):
(-[WKWebViewConfiguration _alternateWebViewForNavigationGestures]):
(-[WKWebViewConfiguration _setAlternateWebViewForNavigationGestures:]):
(-[WKWebViewConfiguration _respectsImageOrientation]):
(-[WKWebViewConfiguration _setRespectsImageOrientation:]):
(-[WKWebViewConfiguration _setAttachmentFileWrapperClass:]):
(-[WKWebViewConfiguration _websiteDataStoreIfExists]):
(-[WKWebViewConfiguration _setCORSDisablingPatterns:]):
(-[WKWebViewConfiguration _maskedURLSchemes]):
(-[WKWebViewConfiguration _setMaskedURLSchemes:]):
(-[WKWebViewConfiguration _setLoadsFromNetwork:]):
(-[WKWebViewConfiguration _setAllowedNetworkHosts:]):
(-[WKWebViewConfiguration _applicationManifest]):
(-[WKWebViewConfiguration _setApplicationManifest:]):
(-[WKWebViewConfiguration _applePayEnabled]):
(-[WKWebViewConfiguration _setApplePayEnabled:]):
(-[WKWebViewConfiguration _setAdditionalSupportedImageTypes:]):
(-[WKWebViewConfiguration _delaysWebProcessLaunchUntilFirstLoad]):
(-[WKWebViewConfiguration _setDelaysWebProcessLaunchUntilFirstLoad:]):
(-[WKWebViewConfiguration _setShouldRelaxThirdPartyCookieBlocking:]):
* Source/WebKit/UIProcess/Cocoa/WebURLSchemeHandlerCocoa.h:
(isType):
* Source/WebKit/UIProcess/WebURLSchemeHandler.h:
(WebKit::WebURLSchemeHandler::isWebURLSchemeHandlerCocoa const):

Canonical link: <a href="https://commits.webkit.org/290789@main">https://commits.webkit.org/290789@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd2a552b82bc25c6d6a4e572358b25152de01572

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91082 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10625 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/117 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96090 "") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41852 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11034 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18945 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/96090 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27527 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94083 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8406 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82534 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/96090 "") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8177 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40982 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78499 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/115 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98068 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18289 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79003 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18548 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78356 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78203 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22712 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/71 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11498 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14381 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18290 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23622 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18016 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21477 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19795 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -120,11 +120,16 @@ - (void)dealloc
     [super dealloc];
 }
 
+- (Ref<API::PageConfiguration>)_protectedPageConfiguration
+{
+    return *_pageConfiguration;
+}
```
Since this is for use inside the class and the source file I might have used a function instead of a method.

cdumez replied:
I wasn't sure how to create a `protectedFoo()` function for an Objective C object. I asked @ddkilzer for help and he recommended adding a method like this. Personally, I'd rather write C++ as much as possible so if there is a way to use a C++ function, that'd be great.

rniwa replied:
FWIW, the static analyzer recognizes these as protectedFoo functions.

darinadler replied:
I just realized that maybe you can't if the field is private so it might not work. But I'd just write a free function like:
```
static Ref<API::PageConfiguration> protectedConfiguration(WKWebViewConfiguration *configuration)
{
    return *configuration->_pageConfiguration;
}
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -120,11 +120,16 @@ - (void)dealloc
         return;
 
-    _pageConfiguration->API::PageConfiguration::~PageConfiguration();
+    [self _protectedPageConfiguration]->API::PageConfiguration::~PageConfiguration();
```
Dot-syntax would be nicer (applies to all other place in this PR).
```suggestion
    self._protectedPageConfiguration->API::PageConfiguration::~PageConfiguration();
```

cdumez replied:
I thought Safari liked the dot-syntax better and WebKit preferred the brackets but it's possible I remember wrong. I don't write a lot of ObjC. I know the bracket syntax looks much better when using RetainPtr but for `self` I guess I don't see a strong reason to not use the dot-syntax.

cdumez committed:
Address safer C++ static analysis warnings in WKWebViewConfiguration
https://bugs.webkit.org/show_bug.cgi?id=288040

Reviewed by Timothy Hatcher and Darin Adler.

* Source/WebKit/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration dealloc]):
(-[WKWebViewConfiguration _protectedPageConfiguration]):
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration processPool]):
(-[WKWebViewConfiguration setProcessPool:]):
(-[WKWebViewConfiguration preferences]):
(-[WKWebViewConfiguration setPreferences:]):
(-[WKWebViewConfiguration userContentController]):
(-[WKWebViewConfiguration setUserContentController:]):
(-[WKWebViewConfiguration _requiredWebExtensionBaseURL]):
(-[WKWebViewConfiguration _setRequiredWebExtensionBaseURL:]):
(-[WKWebViewConfiguration _strongWebExtensionController]):
(-[WKWebViewConfiguration _weakWebExtensionController]):
(-[WKWebViewConfiguration _setWeakWebExtensionController:]):
(-[WKWebViewConfiguration setWebExtensionController:]):
(-[WKWebViewConfiguration websiteDataStore]):
(-[WKWebViewConfiguration setWebsiteDataStore:]):
(-[WKWebViewConfiguration defaultWebpagePreferences]):
(-[WKWebViewConfiguration setDefaultWebpagePreferences:]):
(-[WKWebViewConfiguration _visitedLinkStore]):
(-[WKWebViewConfiguration _setVisitedLinkStore:]):
(-[WKWebViewConfiguration setURLSchemeHandler:forURLScheme:]):
(-[WKWebViewConfiguration urlSchemeHandlerForURLScheme:]):
(-[WKWebViewConfiguration _relatedWebView]):
(-[WKWebViewConfiguration _webViewToCloneSessionStorageFrom]):
(-[WKWebViewConfiguration _setWebViewToCloneSessionStorageFrom:]):
(-[WKWebViewConfiguration _alternateWebViewForNavigationGestures]):
(-[WKWebViewConfiguration _setAlternateWebViewForNavigationGestures:]):
(-[WKWebViewConfiguration _respectsImageOrientation]):
(-[WKWebViewConfiguration _setRespectsImageOrientation:]):
(-[WKWebViewConfiguration _setAttachmentFileWrapperClass:]):
(-[WKWebViewConfiguration _websiteDataStoreIfExists]):
(-[WKWebViewConfiguration _setCORSDisablingPatterns:]):
(-[WKWebViewConfiguration _maskedURLSchemes]):
(-[WKWebViewConfiguration _setMaskedURLSchemes:]):
(-[WKWebViewConfiguration _setLoadsFromNetwork:]):
(-[WKWebViewConfiguration _setAllowedNetworkHosts:]):
(-[WKWebViewConfiguration _applicationManifest]):
(-[WKWebViewConfiguration _setApplicationManifest:]):
(-[WKWebViewConfiguration _applePayEnabled]):
(-[WKWebViewConfiguration _setApplePayEnabled:]):
(-[WKWebViewConfiguration _setAdditionalSupportedImageTypes:]):
(-[WKWebViewConfiguration _delaysWebProcessLaunchUntilFirstLoad]):
(-[WKWebViewConfiguration _setDelaysWebProcessLaunchUntilFirstLoad:]):
(-[WKWebViewConfiguration _setShouldRelaxThirdPartyCookieBlocking:]):
* Source/WebKit/UIProcess/Cocoa/WebURLSchemeHandlerCocoa.h:
(isType):
* Source/WebKit/UIProcess/WebURLSchemeHandler.h:
(WebKit::WebURLSchemeHandler::isWebURLSchemeHandlerCocoa const):

Canonical link: https://commits.webkit.org/290789@main
